### PR TITLE
Sam 1869 battery adjust losses

### PR DESCRIPTION
**Pull request by brtietz**
Add adjust/availability losses to batteries. 

These losses affect both power and capacity, such that when a loss kicks in the battery has x% of each at that time. We assume that the cells have to be discharged to work on them safely, so x% is removed from state of charge and needs to be returned once the loss is over.

Note that this can lead to SOC less than the minimum SOC on the dispatch page. For example, a battery with a 10% minimum SOC and a 50% availability loss could go down to 5%.

Pairs with SAM pull request: https://github.com/NREL/SAM/pull/1902

Closes https://github.com/NREL/SAM/issues/1869

### CODE REVIEW

File: shared/lib_battery_dispatch.cpp [hunk 1]
```diff
@@ -394,17 +399,22 @@ bool dispatch_t::restrict_power(double& I)
         }
         else
         {
-            if (std::abs(powerBattery) > m_batteryPower->powerBatteryDischargeMaxDC * (1 + low_tolerance))
+            double max_discharge_dc = (1 - m_batteryPower->adjustLosses) * m_batteryPower->powerBatteryDischargeMaxDC;
+            double max_discharge_ac = (1 - m_batteryPower->adjustLosses) * m_batteryPower->powerBatteryDischargeMaxAC;
```
[the remark this answers is not in the document]
Comment (brtietz) [reply shared/lib_battery_dispatch.cpp]: Good call - making those variables private revealed several locations I'd missed. Addressed in latest push.

**Comment by brtietz** [comment]
> I'm getting a simulation error now with hourly simulations when the time series battery availability losses are in Hourly mode. This is for both PV-Battery and PVWatts-Battery configurations.
> 
> ```
> exec fail(battery): failed to setup battery adjustment factors: Time series availability losses must have the same time step as the simulation time step unless losses are daily, weekly, monthly, annual, or single value.
> Simulation pvsamv1 failed :exec fail(battery): failed to setup battery adjustment factors: Time series availability losses must have the same time step as the simulation time step unless losses are daily, weekly, monthly, annual, or single value.
> monthly_grid_to_batt not assigned
> Simulation singleowner failed :monthly_grid_to_batt not assigned
> ```

Ugh, thanks for catching that. Fortunately it was an easy fix!

**Comment by cpaulgilman** [comment]
> > I'm getting a simulation error now with hourly simulations when the time series battery availability losses are in Hourly mode. This is for both PV-Battery and PVWatts-Battery configurations.
> > ```
> > exec fail(battery): failed to setup battery adjustment factors: Time series availability losses must have the same time step as the simulation time step unless losses are daily, weekly, monthly, annual, or single value.
> > Simulation pvsamv1 failed :exec fail(battery): failed to setup battery adjustment factors: Time series availability losses must have the same time step as the simulation time step unless losses are daily, weekly, monthly, annual, or single value.
> > monthly_grid_to_batt not assigned
> > Simulation singleowner failed :monthly_grid_to_batt not assigned
> > ```
> 
> Ugh, thanks for catching that. Fortunately it was an easy fix!

Hourly time series losses are  working now for PV-Battery (I tested FOM and BTM), but I'm still getting the simulation error for PVWatts-Battery.

**Comment by brtietz** [comment]
> Hourly time series losses are working now for PV-Battery (I tested FOM and BTM), but I'm still getting the simulation error for PVWatts-Battery.

Thanks for your patience while testing this. This should be fixed - it required changes to both ssc and SAM (lk to pass the right analysis period).

**Comment by cpaulgilman** [comment]
> > Hourly time series losses are working now for PV-Battery (I tested FOM and BTM), but I'm still getting the simulation error for PVWatts-Battery.
> 
> Thanks for your patience while testing this. This should be fixed - it required changes to both ssc and SAM (lk to pass the right analysis period).

PVWatts-Battery is working correctly now. Thanks for the fixes!

**Review by cpaulgilman — APPROVED** [review]

**Review by dguittet — APPROVED** [review]
Thanks!!

**Pull request merged**